utils/zerovec/src/ule/mod.rs

```diff
@@ -19,7 +19,20 @@ pub use plain::PlainOldULE;
 ///
 /// # Safety
 ///
-/// See the safety invariant documented on [`Self::from_byte_slice_unchecked()`] to implement this trait.
+/// There must be no padding bytes involved in this type: [`Self::as_byte_slice()`] MUST return
+/// a slice of initialized bytes provided that `Self` is initialized.
+///
+/// This method _must_ be implemented to return the same result as [`ULE::parse_byte_slice()`].
+///
+/// [`ULE::as_byte_slice()`] should return a slice that is the in-memory representation of `Self`,
+/// i.e. it should be just a pointer cast, and `mem::size_of_val(self) == mem::size_of_val(self.as_byte_slice())`=
+///
+/// # Equality invariant
+///
+/// A non-safety invariant is that if `Self` implements `PartialEq`, it *must* be logically equivalent to
+/// byte equality on `.as_byte_slice()`. Failure to follow this invariant will not cause undefined
+/// behavior, but may cause problems in the `PartialEq` implementations of `ZeroVec` and `VarZeroVec`,
+/// as well as the predictable operation of `ZeroMap`
```

Comment:
Suggestion, here and below:

```suggestion
/// A non-safety invariant is that if `Self` implements `PartialEq`, it *must* be logically equivalent to
/// byte equality on `.as_byte_slice()`. If `PartialEq` does not imply byte-for-byte equality, then
/// `.parse_byte_slice()` should return an error code for any values that are not in canonical form.
///
/// Failure to follow this invariant will cause surprising behavior in `PartialEq`, which may result in
/// unpredictable operations on `ZeroVec`, `VarZeroVec`, and `ZeroMap`.
```